### ISSUE TITLE
Java 8 File.renameTo() fails without error message

### ISSUE THREAD

**Issue by ralfhauser** [systemd/systemd#16899]
Since the upgrade to Debian 10 with tomcat9 (previously Debian 9 with tomcat8) we have problems with moving directories via File.renameTo or rather the underlying java.io.UnixFileSystem.rename.
The reason for this are the sandbox features of systemd which tomcat is using (https://salsa.debian.org/java-team/tomcat9/blob/master/debian/README.Debian).
By default tomcat9 only has access to the following directories (according to /lib/systemd/system/tomcat9.service):
ReadWritePaths=/etc/tomcat9/Catalina/
ReadWritePaths=/var/lib/tomcat9/webapps/
ReadWritePaths=/var/log/tomcat9/

But we would have to move a file from /opt/x/ to /opt/y/. For this we have already added both directories to the ReadWritePaths option.
Writing new files to /opt/y/ works without problems. But moving files from /opt/x/ to /opt/y/ does not work.
If we deactivate the sandbox-features everything works as expected.

/lib/systemd/system/tomcat9.service contains:
[…]
User=tomcat
Group=tomcat
PrivateTmp=yes
AmbientCapabilities=CAP_NET_BIND_SERVICE
NoNewPrivileges=true
CacheDirectory=tomcat9
CacheDirectoryMode=750
ProtectSystem=strict
ReadWritePaths=/etc/tomcat9/Catalina/
ReadWritePaths=/var/lib/tomcat9/webapps/
ReadWritePaths=/var/log/tomcat9/

/etc/systemd/system/tomcat9.service.d/override.conf contains:
[Service]
ReadWritePaths=/opt/x/
ReadWritePaths=/opt/y/

The target directory belongs to another user and tomcat is member of its group (drwxrwxr-x).
For new file writing, it doesn't seem to matter whether the directories belong to user tomcat or another user.
But moving with renameTo fails :(

Any hint?

**Comment by ralfhauser** [comment]
Possibly, the problem is related to java/tomcat as the following test scenario shows:

<<[Unit]
Description=Test service for SystemDs sandboxing features.

[Service]
Type=simple
ExecStart=/bin/bash /usr/libexec/tomcat9/systemd_sandbox_test.sh

# Security
User=tomcat
Group=tomcat

PrivateTmp=yes
ProtectSystem=strict
AmbientCapabilities=CAP_NET_BIND_SERVICE
NoNewPrivileges=true
CacheDirectory=tomcat9
CacheDirectoryMode=750
ReadWritePaths=/opt/sedexclient/
ReadWritePaths=/opt/sedexclient/outbox_temp/
ReadWritePaths=/opt/sedexclient/5.3.1/interface/outbox/

[Install]
WantedBy=multi-user.target


So a Bash-Script, that runs as User tomcat with the same Sandboxing-Settings as the tomcat9 .
The Script is the following:

#!/bin/bash
touch /opt/sedexclient/outbox_temp/testfile
mv /opt/sedexclient/outbox_temp/testfile /opt/sedexclient/5.3.1/interface/outbox/

If I start the Service a file is correctly written to /opt/sedexclient/outbox_temp/testfile and moved without error to /opt/sedexclient/5.3.1/interface/outbox/

Upon removal of the sedexclient-Directories from ReadWritePath , the move results in:
Aug 28 11:18:18 sec-dev bash[8043]: mv: cannot move '/opt/sedexclient/outbox_temp/testfile' to '/opt/sedexclient/5.3.1/interface/outbox/testfile': Read-only file system

**Comment by ralfhauser** [comment]
one intermediate insight:  No Symlinks
So if the directory is
/opt/5.3.1/y 
  but you access it with a "version"-symbolic link
/opt/current/y 
  the move will fail

**Comment by ralfhauser** [comment]
If instead of /opt/x you take
   /tmp
(which in reality results in something like /tmp/systemd-private-93ef0041d89842789fa050c758d-tomcat9.service-PmEq/tmp)
the move will fail as well. This is surprising to me because tomcat should be able to access the source it wrote also for a move ?

**Comment by poettering** [comment]
[Canned reply follows] When submitting a GitHub issue, please fill in the submission form you find in the issue creation page, i.e. provide the requested details, such as distribution and systemd versions. We ask you these details to keep frustration at a minimum, both for us, and for you. If you don't provide this information it makes things needlessly difficult to help you.

Also, please keep in mind: this is a bug and feature request tracker only. If you have a support question, i.e. wonder how to use a feature, or can't figure out how to do something best, please use a different forum, such as the systemd mailing list. Also, do not file bug reports unless you can reproduce the issue with one of the two most recently released *upstream* version of systemd. For everything older, please contact your downstream distribution instead.

Thank you for understanding!

**Comment by 7thsonch** [comment]
I work together with @ralfhauser on this issue.
Now that we added /opt/sedexclient/ to ReadWritePaths (without /opt/sedexclient/outbox_temp/ and /opt/sedexclient/5.3.1/interface/outbox/) everything seems to work as expected. We actually believe that we have already tried the same config and honestly don't know why it suddenly works. Probably it was a combination of the reordering of the RedWritePaths options.
Anyway, everything works as desired now. If we encounter any further problems we will contact the mailing list, thank you!
